### ISSUE TITLE
Create GetWalks operation

### ISSUE THREAD

**Issue by p013570** [gchq/Gaffer#1301]
This operation would allow users to traverse a Gaffer graph and return the walk from each seed to the destination. The path returned should be a list of the Edges traversed (including properties).

The GetWalks operation could simply have a list of GetElements operations. The operation could then be handled by executing a GetElements operation, adding the results to an in memory set, extracting the the vertex at the other end of the edge, then executing the next operation.

After each stage we should remove Edges from the in memory set that are no longer required to reduce the amount of memory used.

After all GetElements operations have been executed we would then need to build up the list of paths.

The return type could be Iterable<Iterable<Edge>>. For example if you queried for A and B.

[
[A->C, C->D],
[A->C, C-E],
[B->D, D->F]
[B->A, A->C]
]

**Comment by p013570** [comment]
Depends on #1443.

**Comment by gaffer01** [comment]
Looking at the current branch for this, I think more discussion is needed as to what we're trying to achieve and the design.

- The operation should not be called `Path`. It should be called `GetPaths` and should return `Path`s.
- We need a `Path` class that can be used to store a path through the graph. A `Path` is more than just an `Iterable` of `Edge`s as there is a guarantee that one of the vertices in the next edge is the same as one of the vertices in the current `Edge`. The `Path` class should enforce this constraint when it is created.
- Are we going to use the `MatchedVertex` field in a `Path` to identify which end of the edge was either the seed (if it's the first edge) or appeared in the previous edge (if it's not the first edge)? If so the `Path` class should enforce this.
- What `Path`s are valid? Is A->B->A valid? Is A->B->A->C valid? Is A->B->A->B valid?
- A `Path` class could include just `Edge`s, or include both `Entities` and `Edges`. We should allow the user to decide which they want. This would then mean that the `GetPath` operation with only 1 hop could return `Entity`s for the seed vertex, `Edge`s and then `Entity`s for the destination vertex (currently users have to manually query to find the properties of vertices at the other end of a hop).
- Creating multiple `GetElements` operations as the input to the operation gives maximum flexibility to the user but is complicated if all you want to do is run multiple `GetEdges`. There could be another operation called `GetNHopPaths` which did this automatically.

**Comment by p013570** [comment]
Thanks @gaffer01. 

- GetPaths is much better - I've update the issue to refer to GetPaths instead of Path.
- A Path class is a good idea, if it extended something like ArrayList then we could still process it as a List. The Path class would then just override add and addAll to validate the Path is connected.
- Yes we need to use the MatchedVertex to ensure we extract the opposite vertex. I actually added a comment about this a few days ago on the pull request for this issue.
- Configuring the operation with a list of GetElements operations means each hop could be down a different Edge group with different filters. So yes, I think all of those paths should be valid.
- We had planned to allow Entities and Entity filtering in a future issue, to reduce the complexity of this initial issue. Currently we just validate and say no Entities are allow in the Views.
- Agreed it is complex, however the flexibility it is really powerful. Yes, we could extend the GetPaths operation and make simpler versions of it.

**Comment by gaffer01** [comment]
Thanks. The other advantage of having a `Path` object is that it can have methods to return the vertex at "distance" N from the original seed. This avoids the user having to iterate through the list of `Edges` until N have been consumed and then use `MatchedVertex` to find the actual distance N vertex.

It might be worth thinking if there is some object that wraps an `Iterable` of `Paths` that would support methods like `getVerticesAtDistance(n)` - that could then be used to return all vertices at distance N from some seeds.

**Comment by p013570** [comment]
Yes that distance field would be helpful.

The getVerticesAtDistance(n) is something we need to be able to do. However, I had imagined we would use a generic operation to do it, see - #1345.

**Comment by gaffer01** [comment]
We should be careful about whether our Path object is actually a trail or a walk: [https://en.wikipedia.org/wiki/Path_(graph_theory)](https://en.wikipedia.org/wiki/Path_(graph_theory)).

In a directed graph A->B->C there is a path A->B->C but there is no path C<-B<-A. But with the current proposed implementation of this operation, C<-B<-A would be returned as path from the seed C.

We need to:
- Be clear about whether we're creating walks, trails, or paths.
- Be clear whether we're treating the graph as undirected (so that C-B-A is a valid path in the above example) or respecting the edge directions (so that it isn't a valid path).

**Comment by p013570** [comment]
Merged into develop.
